tests/rust/safe_abstraction/mutex_u32.rs

```diff
@@ -0,0 +1,444 @@
+   req SysMutex_share(l, ?P);
+   ens *l |-> ?m &*& SysMutex(m, P);
+
+pred SysMutex::new_ghost_arg(P: pred()) = true;
+pred SysMutex_locked(l: *sys::locks::Mutex, P: pred(); t: thread_id_t);
+/* When we have the whole `SysMutex` chunk it means frame is empty, i.e. no other thread has access to this mutex. */
+lem SysMutex_renew(m: sys::locks::Mutex, Q: pred());
+    req SysMutex(m, ?P) &*& Q();
+    ens SysMutex(m, Q);
+/* With these specification we cannot know if the mutex is free, but in presence of `SysMutex_locked(...) we know it is locked.
+   So the proof sketch for this lemma would be:
+   No other thread has access to this mutex.
+   If
+       - The mutex is free and the resource `P` is protected by mutex; It is in the mutex so to speak. We substitute `P` with `Q`
+       and from this point there will be no `[q]SysMutex(m, P)` anywhere. It means we forget about `P`. The state is just like after
+       a call to `sys::locks::Mutex::new` to protect `Q()`.
+
+       - The mutex is locked it means there should be a `SysMutex_locked(m, P(), ?t, ?q)` and `P()` in resources of some thread(s).
+       Operations in this state:
+           - `lock` gets verified which is fine because this call will not terminate unless someone frees the lock. after that we get
+           `Q() &*& SysMutex_locked(m, Q, ..)` and we are in a consistent state.
+           - `unlock` needs to be called by thread `t` and requires `SysMutex_locked(m, P, ?t, ..) &*& P()`. It will soundly free the
+           mutex but here the P() is leaked again. state from here is consistent.
+           - `drop` will get verified and it is sound because `sys::locks::Mutex` implementation simply leaks a locked mutex.
+*/
+@*/
+
+mod sys {
+    pub mod locks {
+        use std::process::abort;
+        pub struct Mutex {
+            m: *mut u32,
+        }
+
+        impl Mutex {
+            pub unsafe fn new() -> Mutex
+            //@ req SysMutex::new_ghost_arg(?P) &*& P();
+            //@ ens SysMutex(result, P);
+            {
+                abort();
+            }
+
+            // Todo: Use `current_thread` var in SysMutex_locked like threading.h. The SysMutex interface does not need `thread_token` in the contracts.
+            pub unsafe fn lock<'a>(&'a self)
+            //@ req thread_token(?t) &*& [?q]SysMutex_share(self, ?P);
+            //@ ens thread_token(t) &*& [q]SysMutex_share(self, P) &*& SysMutex_locked(self, P, t) &*& P();
+            {
+                abort();
+            }
+
+            pub unsafe fn unlock<'a>(&'a self)
+            //@ req thread_token(?t) &*& SysMutex_locked(self, ?P, t) &*& P();
+            //@ ens thread_token(t);
+            {
+                abort();
+            }
+
+            // Todo: impl Drop for Mutex
+            unsafe fn drop<'a>(&'a mut self)
+            //@ req (*self) |-> ?m &*& SysMutex(m, _);
+            //@ ens (*self) |-> m;
+            {
+                abort();
+            }
```

Comment:
This seems unsound. It seems this allows me to unlock a lock that has been dropped? I think `unlock` needs to require a fraction of SysMutex_share.